server/bin/test_data_importer.py

```diff
@@ -0,0 +1,706 @@
+            elif 'subscriptionSubKey' in pool_data:
+                pool['subscriptionSubKey'] = pool_data['subscriptionSubKey']
+            else:
+                pool['subscriptionSubKey'] = 'master'
+
+            if not 'upstreamPoolId' in pool:
+                pool['upstreamPoolId'] = random_id('upstream')
+
+            url = "owners/%s/pools" % (owner_key)
+
+        return self.post(url, pool).json()
+
+    def list_pools(self, owner_key):
+        return self.get("owners/%s/pools" % (owner_key)).json()
+
+    def create_activation_key(self, owner_key, key_data):
+        key = safe_merge(key_data, {})
+        return self.post("owners/%s/activation_keys" % (owner_key), key).json()
+
+    def add_pool_to_activation_key(self, key_id, pool_id, quantity=None):
+        url = "activation_keys/%s/pools/%s" % (key_id, pool_id)
+        params = {}
+
+        if quantity is not None:
+            params['quantity'] = quantity
+
+        return self.post(url, None, params).json()
+
+    def refresh_pools(self, owner_key, wait_for_completion=True):
+        job_status = self.put("owners/%s/subscriptions" % (owner_key), None, params={ 'lazy_regen': True }).json()
+
+        if job_status and wait_for_completion:
+            finished_states = ['ABORTED', 'FINISHED', 'CANCELED', 'FAILED']
+
+            while job_status['state'] not in finished_states:
+                time.sleep(1)
+                job_status = self.get(job_status['statusPath']).json()
+
+            # Clean up job status
+            # self.delete('jobs', {'id': job_status['id']})
+
+        return job_status
+
+
+
+# Gather and order content in a way that ensures linear creation is safe
+def gather_content(data, hosted_test):
+    content = []
+
+    if hosted_test:
+        # In hosted mode, we don't need per-org product or content definitions, so only
+        # add the global content once
+        if 'content' in data:
+            content.extend([(None, value) for value in data['content']])
+
+        if 'owners' in data:
+            for owner in data['owners']:
+                if 'content' in owner:
+                    content.extend([(None, value) for value in owner['content']])
+
+    else:
+        # In standalone mode, we can only add stuff if we have an org to add it to
+        if 'owners' in data:
+            for owner in data['owners']:
+                owner_key = owner['name']
+
+                if 'content' in data:
+                    content.extend([(owner_key, value) for value in data['content']])
+
+                if 'content' in owner:
+                    content.extend([(owner_key, value) for value in owner['content']])
+
+    return content
+
+# Gather and order products in a way that ensures linear creation is safe
+def gather_products(data, hosted_test):
+    def order_products(owner_key, product_map):
+        ordered = []
+        processed_pids = []
+
+        def traverse_product(product):
```

Comment:
Is it really necessary to have inner method here?

Reply:
The method has no meaning outside of the containing method, it closes on some of the containing method's values (processed_pids), and is recursive in nature. Yes, this could be refactored to not require passing all of that around but that seems rather wasteful given closures are available.